### PR TITLE
CRM-17917 - Multiple fixes for fresh Drupal 8 installation

### PR DESCRIPTION
**Pull request by totten**
See detailed descriptions in commit messages.

---
- [CRM-17917: Add support for Drupal 8](https://issues.civicrm.org/jira/browse/CRM-17917)

### CODE REVIEW

**Comment by totten** [comment]
To setup a new D8 build, I'd suggest at time of writing:

```
civibuild create d8-master --patch https://github.com/civicrm/civicrm-drupal/pull/388 --patch https://github.com/civicrm/civicrm-core/pull/8592
```

**Comment by seamuslee001** [comment]
Just noting to set up a new D8 Build now you just need 

`civibuild create d8-master --patch https://github.com/civicrm/civicrm-drupal/pull/388`

File: src/Controller/CivicrmController.php [hunk 1]
```diff
@@ -29,17 +29,9 @@ public static function create(ContainerInterface $container) {
     );
   }
 
-  public function main($args, $one, $two, $three, $four, $five) {
```
Comment (seamuslee001) [inline src/Controller/CivicrmController.php:32]: @totten After i reverted this and applied a small patch to the crmNavigationMenu smarty function i got the civicrm menu reliably loading

Comment (totten) [reply src/Controller/CivicrmController.php:32]: @seamuslee001 , clarification: was that reverting a specific line or all of 495e085ddde163e1078865a17f60d07127bd7d51?

Comment (totten) [reply src/Controller/CivicrmController.php:32]: Also: is the D8 build 8.0, 8.1, or something else? (See `core/CHANGELOG.txt`.) IIRC, I needed the patch in 8.1 to get any routes working...

Comment (seamuslee001) [reply src/Controller/CivicrmController.php:32]: @totten yes and i built this off 8.2 using the command above.

Comment (seamuslee001) [reply src/Controller/CivicrmController.php:32]: clarification it was reverting your whole commit 495e085.

**Comment by eileenmcnaughton** [comment]
@totten any reason not to merge this? My position on 8.x at the moment is that we should merge a bit more agressively if it is to help get it going

**Comment by seamuslee001** [comment]
@eileenmcnaughton The only issue is that i built off 8.2 using buildkit and had to revert one of Tim's commits to make things work right.

**Comment by eileenmcnaughton** [comment]
one of the commits in this PR? Not a good advertisement for it then...

**Comment by jackrabbithanna** [comment]
@seamuslee001 Can you tell me what "crmNavigationMenu smarty function " you had to patch, and how to get this to work?

**Comment by jackrabbithanna** [comment]
@totten , @seamuslee001  , @eileenmcnaughton 
Tested these changes...
I found that totten's patches to src/Routing/Route.php and src/Controller/CivicrmController.php
are necessary, and work. 
the changes in civicrm.install could be better, I like this PR here for that:
https://github.com/civicrm/civicrm-drupal/pull/391 are what should be

Successfully installed CiviCRM into Drupal 8 without civibuild with the changes in the PR and the one mentioned previously.

Drupal 8.2.5
Libraries 8.x-3.x-dev
Jan 4 daily tarball of 4.7

Fixing this bit is good, and should be committed....
I have no CiviCRM menu up top on Civi pages, but the paths themselves work.....that should be a separate issue/PR

**Comment by seamuslee001** [comment]
Thanks Mark, THe test that i needed was that the navigation menu url (that gets loaded by ajax) wasn't quite working as there was an // in instead of a language so i got a patch into core already which sets it to be en_US if not set otherwise. If you have tested and it works for you I'm happy for this to go through

**Comment by jackrabbithanna** [comment]
Ok I needed the changes in this PR to just get any CiviCRM page to load....
I still don't see a menu....
In the watchdog log , I'm seeing page not found errors having to do with the ajax path for the menu.....
So maybe what we need here is a tweak to totten's commit....

the page not found is for this path: (I've got the site in localhost, in a subdirectory of htdocs, no vhost entry, htdocs/drupal825)
http://localhost/drupal825/civicrm/ajax/menujs/202/en_US/1/NFmjWSIO

What should the path to the ajax menu callback be?

**Comment by totten** [comment]
Thanks for the feedback @jackrabbithanna. 

On my local D7, the path looks very similar -- in HTML, it's `/civicrm/ajax/menujs/203/en_US/1/cIVjCDYN` (which resolves to `http://dmaster.l/civicrm/ajax/menujs/203/en_US/1/cIVjCDYN`).

The route `civicrm/ajax/menujs` has an uncommon property where it actually does take advantage of multiple slashes of subordinate paths. (You can see the route is actually defined as [`civicrm/ajax/menujs`](https://github.com/civicrm/civicrm-core/blob/master/CRM/Core/xml/Menu/Admin.xml#L634).)

This suggests that the option `'extra' => '.*',`  isn't having the intended effect (matching the route to all subpaths)?

See also: http://symfony.com/doc/current/routing/slash_in_parameter.html

**Comment by colemanw** [comment]
Merging as most comments suggest this is an improvement on the whole.

**Comment by jackrabbithanna** [comment]
I'll keep pushing on this in a new PR to get the menu going

**Pull request merged**